libbeat/reader/parser/parser.go

```diff
@@ -157,6 +165,14 @@ func (c *Config) Create(in reader.Reader) Parser {
 				return p
 			}
 			p = readjson.NewContainerParser(p, &config)
+		case "syslog":
+			config := syslog.DefaultConfig()
+			cfg := ns.Config()
+			err := cfg.Unpack(&config)
+			if err != nil {
+				return p
+			}
+			p = syslog.NewParser(p, &config)
```

Comment:
I think we need a test to verify parsers are working as expected.

Also I would like to see a test demonstrating one of the use cases which is syslog messages containing multiline data like stack traces. This would only be realistic for a use case where you had something like rsyslog de-muxing the incoming streams to separate files.

```
<165>1 2003-08-24T05:14:15.000003-07:00 192.0.2.1 myproc 8710 - - [beat-logstash-some-name-832-2015.11.28] IndexNotFoundException[no such index]
<165>1 2003-08-24T05:14:15.000003-07:00 192.0.2.1 myproc 8710 - -     at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.resolve(IndexNameExpressionResolver.java:566)
<165>1 2003-08-24T05:14:15.000003-07:00 192.0.2.1 myproc 8710 - -     at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:133)
<165>1 2003-08-24T05:14:15.000003-07:00 192.0.2.1 myproc 8710 - -     at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:77)
<165>1 2003-08-24T05:14:15.000003-07:00 192.0.2.1 myproc 8710 - -     at org.elasticsearch.action.admin.indices.delete.TransportDeleteIndexAction.checkBlock(TransportDeleteIndexAction.java:75)
```

Reply:
I added a couple of tests for parser, including a newline example. I'm not sure how we are going to combine messages like the ones you have listed above. I could be wrong, but the RFC doesn't assign provisions for combining messages like that. The only thing it may mention is that it's up to the relay or collector to handle this (rsyslog in this case?). I'm not too familiar with rsyslog's demuxing feature, so if you have any pointers on how that works, I'll see what I can do to support that.

It seems to me though that the multiline parser should handle the heavy lifting here, unless there's something specific in the syslog protocol that we can leverage.

Reply:
Yes, multiline should be able to do it. You would apply `syslog` first and then `multiline` would join the data contained in the `message` part. So the multiline pattern would be based on the syslog message data only and not the header.

My preamable about rsyslog was to highlight that this isn't a realistic use case if you were trying to apply a multiline parser to data coming from more than one network source.

Reply:
I added a test for this scenario. That part worked fine (just had to remove a whitespace trim that I had when setting the message field), but one thing I noticed was that setting event.original is a bit weird. It'll only have the last message that was handled, not the whole set of original messages. That makes sense, though. Perhaps for parsers, it doesn't make sense to set event.original since we can't predict what happens before and after this parser runs.

Here's the event that's emitted from the example above for reference:

```json
{
    ...
    "event": {
        ...
        "original": "<165>1 2003-08-24T05:14:30.000003-07:00 192.0.2.1 myproc 8710 - -     at org.elasticsearch.action.admin.indices.delete.TransportDeleteIndexAction.checkBlock(TransportDeleteIndexAction.java:75)"
    },
    "message": "[beat-logstash-some-name-832-2015.11.28] IndexNotFoundException[no such index]\n    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.resolve(IndexNameExpressionResolver.java:566)\n    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:133)\n    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:77)\n    at org.elasticsearch.action.admin.indices.delete.TransportDeleteIndexAction.checkBlock(TransportDeleteIndexAction.java:75)"
}
```

Reply:
> Perhaps for parsers, it doesn't make sense to set event.original 

I agree. I think it would be hard to get this correct without specification of the expected behavior in the parser interface.

I think I would want event.original to be automatically set once on the way in to the parsers and then any processors that aggregate multiple events should append all `event.original` values into an array.